File: src/models/options.rs
use super::Greeks;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OptionsChainResponse {
    pub status: String,
    pub request_id: String,
    pub results: Vec<OptionContract>,
    pub next_url: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct OptionContract {
    pub details: Option<ContractDetails>,
    pub greeks: Option<Greeks>,
    pub implied_volatility: Option<f64>,
    pub open_interest: Option<i64>,
    pub last_quote: Option<LastQuote>,
    pub last_trade: Option<LastTrade>,
    pub day: Option<DayBar>,
    pub underlying_asset: Option<UnderlyingAsset>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ContractDetails {
    pub ticker: Option<String>,
    pub contract_type: Option<String>,
    pub strike_price: Option<f64>,
    pub expiration_date: Option<String>,
    pub exercise_style: Option<String>,
    pub shares_per_contract: Option<i32>,
}

impl ContractDetails {
    pub fn contract_type_enum(&self) -> Option<ContractType> {
        self.contract_type.as_ref().and_then(|t| match t.as_str() {
            "call" => Some(ContractType::Call),
            "put" => Some(ContractType::Put),
            _ => None,
        })
    }

    pub fn exercise_style_enum(&self) -> Option<ExerciseStyle> {
        self.exercise_style
            .as_ref()
            .and_then(|s| match s.as_str() {
                "american" => Some(ExerciseStyle::American),
                "european" => Some(ExerciseStyle::European),
                "bermudan" => Some(ExerciseStyle::Bermudan),
                _ => None,
            })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ContractType {
    Call,
    Put,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExerciseStyle {
    American,
    European,
    Bermudan,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct LastQuote {
    pub bid: Option<f64>,
    pub ask: Option<f64>,
    pub bid_size: Option<i64>,
    pub ask_size: Option<i64>,
}

impl LastQuote {
    pub fn mid_price(&self) -> Option<f64> {
        match (self.bid, self.ask) {
            (Some(bid), Some(ask)) => Some((bid + ask) / 2.0),
            _ => None,
        }
    }

    pub fn spread(&self) -> Option<f64> {
        match (self.bid, self.ask) {
            (Some(bid), Some(ask)) => Some(ask - bid),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct LastTrade {
    pub price: Option<f64>,
    pub size: Option<i64>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct DayBar {
    pub open: Option<f64>,
    pub high: Option<f64>,
    pub low: Option<f64>,
    pub close: Option<f64>,
    pub volume: Option<i64>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UnderlyingAsset {
    pub ticker: Option<String>,
    pub price: Option<f64>,
}

File: backend-go/internal/models/options.go
package models

// OptionsChainResponse represents the response from Massive API
type OptionsChainResponse struct {
	Status    string           `json:"status"`
	RequestID string           `json:"request_id"`
	Results   []OptionContract `json:"results"`
	NextURL   *string          `json:"next_url,omitempty"`
}

// OptionContract represents a single options contract with all market data
type OptionContract struct {
	Details          *ContractDetails `json:"details,omitempty"`
	Greeks           *Greeks          `json:"greeks,omitempty"`
	ImpliedVol       *float64         `json:"implied_volatility,omitempty"`
	OpenInterest     *int64           `json:"open_interest,omitempty"`
	LastQuote        *LastQuote       `json:"last_quote,omitempty"`
	LastTrade        *LastTrade       `json:"last_trade,omitempty"`
	Day              *DayBar          `json:"day,omitempty"`
	UnderlyingAsset  *UnderlyingAsset `json:"underlying_asset,omitempty"`
}

// ContractDetails contains the contract specifications
type ContractDetails struct {
	Ticker            *string  `json:"ticker,omitempty"`
	ContractType      *string  `json:"contract_type,omitempty"` // "call" or "put"
	StrikePrice       *float64 `json:"strike_price,omitempty"`
	ExpirationDate    *string  `json:"expiration_date,omitempty"`
	ExerciseStyle     *string  `json:"exercise_style,omitempty"` // "american", "european", "bermudan"
	SharesPerContract *int     `json:"shares_per_contract,omitempty"`
}

// Greeks contains the option Greeks
type Greeks struct {
	Delta *float64 `json:"delta,omitempty"`
	Gamma *float64 `json:"gamma,omitempty"`
	Theta *float64 `json:"theta,omitempty"`
	Vega  *float64 `json:"vega,omitempty"`
	Rho   *float64 `json:"rho,omitempty"`
}

// LastQuote contains the most recent bid/ask data
type LastQuote struct {
	Bid     *float64 `json:"bid,omitempty"`
	Ask     *float64 `json:"ask,omitempty"`
	BidSize *int64   `json:"bid_size,omitempty"`
	AskSize *int64   `json:"ask_size,omitempty"`
}

// MidPrice calculates the mid-price between bid and ask
func (q *LastQuote) MidPrice() *float64 {
	if q.Bid != nil && q.Ask != nil {
		mid := (*q.Bid + *q.Ask) / 2.0
		return &mid
	}
	return nil
}

// Spread calculates the bid-ask spread
func (q *LastQuote) Spread() *float64 {
	if q.Bid != nil && q.Ask != nil {
		spread := *q.Ask - *q.Bid
		return &spread
	}
	return nil
}

// LastTrade contains the most recent trade data
type LastTrade struct {
	Price *float64 `json:"price,omitempty"`
	Size  *int64   `json:"size,omitempty"`
}

// DayBar contains OHLCV data for the trading day
type DayBar struct {
	Open   *float64 `json:"open,omitempty"`
	High   *float64 `json:"high,omitempty"`
	Low    *float64 `json:"low,omitempty"`
	Close  *float64 `json:"close,omitempty"`
	Volume *int64   `json:"volume,omitempty"`
}

// PercentChange calculates the % change from open to current price
func (d *DayBar) PercentChange(currentPrice float64) *float64 {
	if d.Open != nil && *d.Open > 0 {
		change := ((currentPrice - *d.Open) / *d.Open) * 100.0
		return &change
	}
	return nil
}

// UnderlyingAsset contains information about the underlying stock
type UnderlyingAsset struct {
	Ticker *string  `json:"ticker,omitempty"`
	Price  *float64 `json:"price,omitempty"`
}
